interfaces/prompting/patterns/patterns_test.go

```diff
@@ -864,3 +864,644 @@ func (s *patternsSuite) TestPathPatternMatchesErrors(c *C) {
 	c.Check(err, Equals, doublestar.ErrBadPattern)
 	c.Check(matches, Equals, false)
 }
+
+func (s *patternsSuite) TestHighestPrecedencePattern(c *C) {
+	for i, testCase := range []struct {
+		matchingPath      string
+		patterns          []string
+		highestPrecedence string
+	}{
+		{
+			"/foo",
+			[]string{
+				"/foo",
+			},
+			"/foo",
+		},
+		{
+			"/foo/bar/baz",
+			[]string{
+				"/foo/bar/baz",
+				"/foo/bar/ba?",
+			},
+			"/foo/bar/baz",
+		},
+		{
+			"/foo/bar/baz",
+			[]string{
+				"/foo/bar/b?z",
+				"/foo/bar/baz",
+			},
+			"/foo/bar/baz",
+		},
+		{
+			"/foo/bar/baz",
+			[]string{
+				"/foo/bar/baz*",
+				"/foo/bar/baz",
+			},
+			"/foo/bar/baz",
+		},
+		{
+			"/foo/bar/baz",
+			[]string{
+				"/foo/b?r/baz",
+				"/foo/bar/**",
+			},
+			"/foo/bar/**",
+		},
+		{
+			"/foo/bar/",
+			[]string{
+				"/foo/bar/",
+				"/foo/bar",
```

Comment:
do both actually match /foo/bar/ ?

Reply:
They do using the matching rules we've defined for prompting: https://github.com/snapcore/snapd/pull/13868/files#diff-5be32cc03287c0a9f122784b25e9e2a61a0340e72ed2bd18a849a008cad895bfR109-R144

This may differ some from bash's globstar (the inspiration (rather than AppArmor rules) for the pattern matching system). The idea here is to be pragmatic. Rather than a user needing to always specify `/foo/bar{,/}` at the end of each pattern, we instead let `/foo/bar` match whether `bar` is a file or directory (as this is how the shell generally works), and if they want to restrict access to match only if a path is a directory, then they can specify `/foo/bar/` to do so.